Add `purge-signatures` subcommand

georgepisaltu opened the pull request:
Fixes #33 

This PR adds the `purge-signatures` subcommand. It takes a path to a storage file and lists of block heights for which to purge all signatures or just the required signatures to reach weak but not strict finality.

georgepisaltu committed:
Update dependencies

Signed-off-by: George Pisaltu <georgep@casperlabs.io>

georgepisaltu committed:
Add `purge-signatures` subcommand

Signed-off-by: George Pisaltu <georgep@casperlabs.io>

georgepisaltu committed:
Temporarily allow cargo audit warnings

Signed-off-by: George Pisaltu <georgep@casperlabs.io>

georgepisaltu committed:
Further fix clippy warnings

Signed-off-by: George Pisaltu <georgep@casperlabs.io>

rafal-ch commented on src/subcommands/purge_signatures/signatures.rs:
```diff
@@ -0,0 +1,281 @@
+    }
+    let mut accumulated_sigs: BTreeSet<&PublicKey> = Default::default();
+    let mut accumulated_weight = U512::zero();
+    // Start from the smalles signatures and add them to our pool until weak
```
```suggestion
    // Start from the smallest signatures and add them to our pool until weak
```

georgepisaltu replied:
Done.

rafal-ch commented on src/subcommands/purge_signatures/purge.rs:
```diff
@@ -0,0 +1,266 @@
+
+pub fn purge_signatures<P: AsRef<Path>>(
+    db_path: P,
+    weak_finality_block_list: Vec<u64>,
+    no_finality_block_list: Vec<u64>,
```
A nit:
These two could already be created as `BTreeSet` in `fn run()`, then you can even consider `purge_signatures_for_blocks()` to accept `ExactSizeIterator`.

But that's just a style preference, so feel free to ignore :)

georgepisaltu replied:
Initially I wanted to not do any "computation" logic in `run`, just argument parsing. And a `Vec` is basically just an owned slice, that's what I was going for, but right now it seems a bit weird to have it first collected into a `Vec` and then turned into a `BTreeSet`, so I just did that in `run`. I won't do the `ExactSizeIterator` thing though as I prefer passing in the actual thing. Hope that's ok with you.

rafal-ch reviewed (approved):
No blockers, good job! :+1:

georgepisaltu committed:
Fix typo in signature.rs documentation

Signed-off-by: George Pisaltu <georgep@casperlabs.io>

georgepisaltu committed:
Move list initialization to a higher level

Signed-off-by: George Pisaltu <georgep@casperlabs.io>

georgepisaltu committed:
Add test for missing sigs in database

Signed-off-by: George Pisaltu <georgep@casperlabs.io>

georgepisaltu committed:
Fix purge behavior for missing sigs entry in db

Signed-off-by: George Pisaltu <georgep@casperlabs.io>

Fraser999 commented on src/subcommands/purge_signatures.rs:
```diff
@@ -0,0 +1,125 @@
+                .required_unless_present(NO_FINALITY)
+                .short('w')
+                .long(WEAK_FINALITY)
+                .takes_value(true)
```
You could have probably used `use_value_delimiter(true)` here instead and that would avoid having to manually parse the multiple values.  Not a big deal though.

georgepisaltu replied:
I'd like to not do this right now as I'm more confident with manual parsing. This is something we should do at some point in the future though when we finally add CLI usage tests and/or we end up having another parameter which takes multiple values, like a list. Let me know if that's ok with you.

Fraser999 replied:
Fine by me :)

Fraser999 commented on src/subcommands/purge_signatures/signatures.rs:
```diff
@@ -0,0 +1,281 @@
+    // Store the signature keys sorted by their respective weight.
+    let mut inverse_map: BTreeMap<U512, Vec<&PublicKey>> = BTreeMap::default();
+    for (key, weight) in weights.iter() {
+        match inverse_map.entry(*weight) {
+            Entry::Vacant(vacant_entry) => {
+                vacant_entry.insert(vec![key]);
+            }
+            Entry::Occupied(mut occupied_entry) => occupied_entry.get_mut().push(key),
+        }
```
```suggestion
        inverse_map.entry(*weight).or_default().push(key);
```

georgepisaltu replied:
Done.

Fraser999 commented on src/subcommands/purge_signatures/purge.rs:
```diff
@@ -0,0 +1,274 @@
+                WriteFlags::default(),
+            )?;
+        } else {
+            info!("Couldn't strip signatures for block {block_hash} at height {block_height}");
```
Maybe worth a `warn!` rather than `info!`?

georgepisaltu replied:
Done.

Fraser999 commented on src/subcommands/purge_signatures/tests.rs:
```diff
@@ -0,0 +1,728 @@
+        );
+        assert!(!era_weights.weights_mut().contains_key(&KEYS[0]));
+
+        // Try to update the weights for a inexistent switch block.
```
```suggestion
        // Try to update the weights for a nonexistent switch block.
```

georgepisaltu replied:
Done.

Fraser999 commented on src/subcommands/purge_signatures/tests.rs:
```diff
@@ -0,0 +1,728 @@
+        assert!(block_2_sigs.proofs.contains_key(&KEYS[0]));
+        assert!(!block_2_sigs.proofs.contains_key(&KEYS[1]));
+
+        // Block 2 should be the same as before.
```
```suggestion
        // Block 3 should be the same as before.
```

georgepisaltu replied:
Done.

Fraser999 commented on src/subcommands/purge_signatures/purge.rs:
```diff
@@ -0,0 +1,274 @@
+impl EraWeights {
+    /// Update the internal structure to hold the validator weights for
+    /// the era given as input.
+    pub(crate) fn refresh_weights_for_era<T: Transaction>(
```
I think this is sufficient for now, but we should ticket updating this method (and any other places in db-utils like this) to take validator-changing upgrades into account.  (We handle this via the `GotBlockAfterUpgradeEraValidators` event in casper-node I think).

For now, maybe the tool could print a warning if the switch block used inside here happens to be the final switch block of a given protocol version?  That would require parsing the subsequent block to check for a different version in it.

georgepisaltu replied:
I tried to address this the best I could without starting to query global state for the weights. The tool and the CLI interface is built in such a way that the user knows what blocks they want stripped, so hopefully they don't intentionally choose a switch block for upgrade.

georgepisaltu committed:
Address upgrade switch blocks scenario

Signed-off-by: George Pisaltu <georgep@casperlabs.io>

georgepisaltu committed:
Various documentation and style adjustments

Signed-off-by: George Pisaltu <georgep@casperlabs.io>

georgepisaltu committed:
Add h2 to audit ignore list

The `h2` dependency is out of date. This commit adds this to the
`cargo audit` ignore list because it is not trivial to just update to a
version that is safe because of dependency mismatches.

Signed-off-by: George Pisaltu <georgep@casperlabs.io>

Fraser999 reviewed (approved):
LGTM - nice one!
